### ISSUE TITLE
False negative? for unicorn/consistent-function-scoping

### ISSUE THREAD

**Issue by dashmug** [sindresorhus/eslint-plugin-unicorn#372]
I like this new rule and I've made change to my code. Here's one WIP snippet where it proved to be problematic though.

```ts
type Data<T> = T extends 'error' ? Error : Record<string, unknown> | unknown[]

type Method = 'info' | 'error'

export function createLogger(name: string) {
    // Two lint errors are on the next line.
    const log = <T extends Method>(method: T) => (data: Data<T>) => {
        try {
            // eslint-disable-next-line no-console
            console[method](JSON.stringify({ name, data }))
        } catch (error) {
            console.error(error)
        }
    }

    return {
        info: log('info'),
        error: log('error'),
    }
}
```

```
6:17  error  Move arrow function to the outer scope  unicorn/consistent-function-scoping
6:50  error  Move arrow function to the outer scope  unicorn/consistent-function-scoping
```

Could this be a bug in `@typescript-eslint`?

@typescript-eslint/eslint-plugin@2.3.0
@typescript-eslint/parser@2.3.0
eslint-plugin-unicorn@11.0.0

**Comment by MrHen** [comment]
Yeah, that doesn't look right. I'll try to take a look at it later tonight.

**Comment by buschtoens** [comment]
I have a very similar case.

```ts
test('it works', async function(assert) {
  function assertHeader(assertions) {
    for (const [key, value] of Object.entries(assertions)) {
      assert.strictEqual(
        native[key],
        value,
        `'native.${key}' was updated to '${value}'.`
      );
    }
  }

   // ...
});
```

If I lift the `assert` reference out of the `for` loop, the false negative vanishes.

```diff
test('it works', async function(assert) {
  function assertHeader(assertions) {
+   assert;
    for (const [key, value] of Object.entries(assertions)) {
      assert.strictEqual(
        native[key],
        value,
        `'native.${key}' was updated to '${value}'.`
      );
    }
  }

   // ...
});
```

I did not look at the rule source code yet, but it looks like only the root function body scope is checked for references.

**Comment by fobdy** [comment]
Have the similar issue at line 2 with the following:

```ts
export function a(x: number) {
  const b = (y: number) => (z: number): number => x + y + z;
  return b(1)(2);
}
```

**Comment by lvdang** [comment]
I am seeing this issue in a .js file.

![Screen Shot 2020-10-14 at 2 00 49 PM](https://user-images.githubusercontent.com/7051557/96044913-04fa7100-0e26-11eb-9c7e-0986d6fddec5.png)


"@typescript-eslint/eslint-plugin": "4.4.1",
 "@typescript-eslint/parser": "4.4.1",
 "eslint-plugin-unicorn": "22.0.0",
